### PR TITLE
Update tdex to v1.0.0

### PR DESCRIPTION
**Pull request by altafan**
This updates TDEX from `v0.9.1` to `v1.0.0`.

The compose file defines a brand new `oceand` service required by the new version of `tdexd` v1.0.0. Updated also the dashboard to v1.0.1.

### CODE REVIEW

File: tdex/hooks/pre-start [hunk 1]
```diff
@@ -6,8 +6,9 @@ if [[ -f "${HIDDEN_SERVICE_FILE}" ]]; then
 	exit
 fi
 
-"${UMBREL_ROOT}/scripts/app" compose "${APP_ID}" up --detach tdexd
 "${UMBREL_ROOT}/scripts/app" compose "${APP_ID}" up --detach tor
+"${UMBREL_ROOT}/scripts/app" compose "${APP_ID}" up --detach oceand
+"${UMBREL_ROOT}/scripts/app" compose "${APP_ID}" up --detach tdexd
```
Comment (nmfretz) [inline tdex/hooks/pre-start:11]: I think you can revert this change. The `oceand` container (as well as the dashboard) will be brought up after this script runs. This is just making sure that a tor hidden service is created for the daemon.

File: tdex/docker-compose.yml [hunk 2]
```diff
@@ -6,31 +6,52 @@ services:
...
+    image: ghcr.io/tdex-network/tdexd:v1.0.0@sha256:6e6733015dceeab41db4780b5e99313d5bbbc5dc490800ee3b5a326a8b65b92d
     tty: true
     restart: on-failure
+    depends_on:
+      - oceand
     stop_grace_period: 1m
     volumes:
       - ${APP_DATA_DIR}/tdex-data:/home/tdex/.tdex-daemon
+    ports:
+     - ${APP_TDEX_PORT}:${APP_TDEX_PORT}
```
Comment (nmfretz) [inline tdex/docker-compose.yml:33]: we don't need to map ports here, as Docker's internal DNS resolution will allow the containers in this compose file to connect to each other. Unless you are trying to expose this port on the host so that it is reachable externally?

**Review by nmfretz — COMMENTED** [review]
Thanks for this update @altafan! I have left some suggestions below for you to address before I test.

Also, are you okay if I push a commit directly to your `update-tdex` branch? We need to add a pre-start hook to make sure that users who already have tdex installed receive the new `ocean-data` directory with appropriate ownership when they update.

File: tdex/docker-compose.yml [hunk 3]
```diff
@@ -6,31 +6,52 @@ services:
       APP_HOST: tdex_caddy_1
       APP_PORT: 8080
 
+  oceand:
+    image: ghcr.io/vulpemventures/oceand:v0.1.17@sha256:9a7e1b8df4d7907ad66e3475c5a5f357d73c49df96214c00eba121f2aea9a50d
```
[the remark this answers is not in the document]
Comment (altafan) [reply tdex/docker-compose.yml:10]: I'm sorry, where did you get this digest from? I can't find it [here](https://github.com/vulpemventures/ocean/pkgs/container/oceand)

File: tdex/docker-compose.yml (hunk 2 above)
Comment (altafan) [reply tdex/docker-compose.yml:33]: Yeah the idea is to make this service accessible from the outside world.

File: tdex/hooks/pre-start (hunk 1 above)
Comment (altafan) [reply tdex/hooks/pre-start:11]: I made the change because `tdexd` requires an already running `oceand` to start correctly. 
In light of this, do you think I should still revert the changes?

**Comment by altafan** [comment]
> Thanks for this update @altafan! I have left some suggestions below for you to address before I test.
>
> Also, are you okay if I push a commit directly to your update-tdex branch? We need to add a pre-start hook to make sure that users who already have tdex installed receive the new ocean-data directory with appropriate ownership when they update.

Thank you @nmfretz for the review!

I guess I have to give you access to my repo in order to directly push changes? Can you wait until all required changes are fixed?

File: tdex/docker-compose.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (nmfretz) [reply tdex/docker-compose.yml:10]: When you click on the `...` next to `Digest` it shows you the multi-arch digest:

image

This works because the v0.1.17 image is multi-arch as seen here (both amd64 and arm64):

image

Another way you can get the correct digest for multi-arch images is to run `docker pull <image><tag>` and the correct digest will be shown in the output. For example:

```
$ sudo docker pull ghcr.io/vulpemventures/oceand:v0.1.17
v0.1.17: Pulling from vulpemventures/oceand
...
Digest: sha256:37b813b43b068e32d918b9f42539c33287308cc81d00909c742b7164dfefc0c4
Status: Downloaded newer image for ghcr.io/vulpemventures/oceand:v0.1.17
ghcr.io/vulpemventures/oceand:v0.1.17
```

File: tdex/hooks/pre-start (hunk 1 above)
Comment (nmfretz) [reply tdex/hooks/pre-start:11]: ah I see, sorry for misunderstanding. I see now that you have tdex depending on oceand in the compose file. Great work catching this!

**Comment by nmfretz** [comment]
> I guess I have to give you access to my repo in order to directly push changes? Can you wait until all required changes are fixed?

Actually, here's what I think will be needed. You can add this above the Tor Hidden Service logic in the existing `pre-start`. 

```bash
set -euo pipefail

APP_DATA_DIR="$(readlink -f $(dirname "${BASH_SOURCE[0]}")/..)"
OCEAN_DATA_DIR="${APP_DATA_DIR}/ocean-data"

[ ! -d "${OCEAN_DATA_DIR}" ] && mkdir -p "${OCEAN_DATA_DIR}" && chown 1000:1000 "${OCEAN_DATA_DIR}"
```

File: tdex/docker-compose.yml (hunk 2 above)
Comment (nmfretz) [reply tdex/docker-compose.yml:33]: ah okay, so in that case we need to change `${APP_TDEX_PORT}` to something other than `9090` because this port has been taken by Urbit Bitcoin Connector, meaning if a user is running both TDEX and Urbit Bitcoin Connector there will be a port clash on the host and one of the apps will break:
https://github.com/getumbrel/umbrel-apps/blob/5ec41a7ae69eb202cf43533d8749a105a4444477/urbit-bitcoin-connector/umbrel-app.yml#L29

Comment (nmfretz) [reply tdex/docker-compose.yml:33]: ah okay, so in that case we need to change `${APP_TDEX_PORT}` to something other than `9090` because this port has been taken by Urbit Bitcoin Connector, meaning if a user is running both TDEX and Urbit Bitcoin Connector there will be a port clash on the host and one of the apps will break:
https://github.com/getumbrel/umbrel-apps/blob/5ec41a7ae69eb202cf43533d8749a105a4444477/urbit-bitcoin-connector/umbrel-app.yml#L29

Comment (altafan) [reply tdex/docker-compose.yml:33]: it shouldn't be a problem at all, do you suggest any port number?

Comment (altafan) [reply tdex/docker-compose.yml:33]: what about `9000`?

File: tdex/docker-compose.yml [hunk 4]
```diff
@@ -6,31 +6,52 @@ services:
...
       # reachable remotely by other dashboards via Tor
       TDEX_CONNECT_ADDR: "${APP_TDEX_DAEMON_HIDDEN_SERVICE}"
       TDEX_CONNECT_PROTO: http
 
   dashboard:
-    image: ghcr.io/tdex-network/dashboard:v0.1.55@sha256:dfa7cdb75117c97e671285fc9754a6cabafc7f74760c314e6625c33067509beb
+    image: ghcr.io/tdex-network/dashboard:v1.0.1@sha256:a99b72c76190453b74f16e3f4961f99a3d84ba7ec4315bd913247f7d49bff10f
     restart: on-failure
+    ports:
+      - 8080:8080
```
[the remark this answers is not in the document]
Comment (altafan) [reply tdex/docker-compose.yml]: done.

File: tdex/docker-compose.yml [hunk 5]
```diff
@@ -6,31 +6,52 @@ services:
...
+      TDEX_WALLET_ADDR: oceand:18000
       # these vars are used to generate the connect URL
       # we assume to always show the external endpoint
       # reachable remotely by other dashboards via Tor
       TDEX_CONNECT_ADDR: "${APP_TDEX_DAEMON_HIDDEN_SERVICE}"
       TDEX_CONNECT_PROTO: http
 
   dashboard:
-    image: ghcr.io/tdex-network/dashboard:v0.1.55@sha256:dfa7cdb75117c97e671285fc9754a6cabafc7f74760c314e6625c33067509beb
+    image: ghcr.io/tdex-network/dashboard:v1.0.1@sha256:a99b72c76190453b74f16e3f4961f99a3d84ba7ec4315bd913247f7d49bff10f
```
[the remark this answers is not in the document]
Comment (altafan) [reply tdex/docker-compose.yml]: done (we released v1.0.2 in the meantime).

File: tdex/docker-compose.yml [hunk 6]
```diff
@@ -6,31 +6,52 @@ services:
...
+    environment:
+      OCEAN_LOG_LEVEL: 5
+      OCEAN_DB_TYPE: "badger"
+      OCEAN_NO_TLS: "true"
+      OCEAN_NO_PROFILER: "true"
+      OCEAN_UTXO_EXPIRY_DURATION_IN_SECONDS: 240
+
   tdexd:
-    image: ghcr.io/tdex-network/tdexd:v0.9.1@sha256:5078897b9a505fc3dfac5b3d51537150941c075f5279ae81de1c6b1c21112544
+    image: ghcr.io/tdex-network/tdexd:v1.0.0@sha256:6e6733015dceeab41db4780b5e99313d5bbbc5dc490800ee3b5a326a8b65b92d
```
[the remark this answers is not in the document]
Comment (altafan) [reply tdex/docker-compose.yml]: done (we released v1.0.1 in the meantime).

File: tdex/docker-compose.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (altafan) [reply tdex/docker-compose.yml]: done.

File: tdex/docker-compose.yml (hunk 2 above)
Comment (ParthJadhav) [reply tdex/docker-compose.yml:33]: Hi @altafan , port 9000 is currently occupied by portainer. To avoid conflicts, please consider using 9002, which is the closest available option. Before selecting a port, it's a good practice to search within this repository to ensure it doesn't overlap with existing port assignments.

**Comment by nmfretz** [comment]
Hey @altafan can you please update the exposed daemon port to 9002: https://github.com/getumbrel/umbrel-apps/pull/754#discussion_r1335746250

Also, what would be the best way for us to test external connection in the tdex ecosystem?

**Comment by altafan** [comment]
> Also, what would be the best way for us to test external connection in the tdex ecosystem?

@nmfretz I made a test on a remote host, and once I installed the tdex app, I just went to `<ip>:9090` to see a page similar to the the following one:

**Comment by tiero** [comment]
Any news? @nmfretz

**Comment by nmfretz** [comment]
oh sorry, @tiero and @altafan can you guys push the quick fix noted here: https://github.com/getumbrel/umbrel-apps/pull/754#discussion_r1326207517. Port 9090 clashes with an existing app. 9092 is free for example.

Also, since this is an update, we can now add release notes to the umbrel-app.yml file. See Lightning app as an example:
https://github.com/getumbrel/umbrel-apps/blob/776587169cba9e345fcbc9e54f82675fd10e9eb2/lightning/umbrel-app.yml#L24-L28

**Comment by altafan** [comment]
@nmfretz i pushed the changes. Now the port is `9092`. I also followed your suggestion and added some release notes, thanks 🙏.

**Comment by nmfretz** [comment]
Thanks @altafan. I have made a few changes and we are nearly there:
- in your recent commit [19e9840](https://github.com/getumbrel/umbrel-apps/pull/754/commits/19e9840d65ab3222c27268ebb3886ad182783a65), you changed the port in the manifest to 9092. This port is actually for your **dashboard UI**, so should stay as 9094. I have changed that here [a5bac52](https://github.com/getumbrel/umbrel-apps/pull/754/commits/a5bac524cd2a6ae250251cd1d478a2fd866b390b)
- updated the Caddyfile so that new installs start with the correct tdexd port (9092) [2f61879](https://github.com/getumbrel/umbrel-apps/pull/754/commits/2f618791737b9e0b8056a91e21b96e6639f0020d)
- added logic to the pre-start hook so that existing installs that are updating to this new version receive the change to the caddyfile [f566598](https://github.com/getumbrel/umbrel-apps/pull/754/commits/f5665983f8a1029cb1792693850c529df560f472)

I tested on a new install and it works well 👌. However, something breaks in the tdexd container when updating the app from the previous version. Here is the error I get from the tdexd container

```
$ sudo docker logs tdex_tdexd_1
INFO[0000] All 0 tables opened in 0s
INFO[0000] Discard stats nextEmptySlot: 0
INFO[0000] Set nextTxnTs to 0
INFO[0000] Deleting empty file: /home/tdex/.tdex-daemon/db/feeder/000007.vlog
INFO[0000] All 0 tables opened in 0s
INFO[0000] Discard stats nextEmptySlot: 0
INFO[0000] Set nextTxnTs to 0
INFO[0000] Deleting empty file: /home/tdex/.tdex-daemon/db/markets/000007.vlog
FATA[0000] failed to initialize grpc service             error="invalid opts: invalid app config: opening prices db: manifest has unsupported version: 7 (we support 8).\nPlease see https://github.com/dgraph-io/badger/blob/master/README.md#i-see-manifest-has-unsupported-version-x-we-support-y-error on how to fix this."
```

Could you please take a look into this? Once this is fixed we can go live!

**Comment by altafan** [comment]
@nmfretz sorry for the late reply

> However, something breaks in the tdexd container when updating the app from the previous version. Here is the error I get from the tdexd container

The reason is that the database of the prev version of tdexd must be migrated in order to be compatible with the new version.
The migration tool is packed within the official docker image, do you think it's possible to setup a migration script that runs it if needed to solve the issue?

**Comment by altafan** [comment]
@nmfretz let's go for the easiest way and eventually remove the datadir if existing. Can you tell me if I can add this conditional operation to some hook script?

**Comment by nmfretz** [comment]
@tiero @altafan sorry for the delay.

> Otherwise, it could just be enough to delete the existing datadir, the daemon will create a new one at first startup and the user will restore the wallet, which seems an easier solution.

For this option we'd run the risk of users clicking update without reading the release notes, so there could be a subset of users who update and don't have (or have lost) their recovery phrase, which would be pretty bad I think.

Additionally we'd have to make sure that this database deletion logic doesn't run on subsequent app starts/restarts/updates, which is totally possible but adds complexity.

> The migration tool is packed within the official docker image, do you think it's possible to setup a migration script that runs it if needed to solve the issue?

Yes, we can do this with a pre-start script. The benefit here is that we users won't accidently lose access to their funds.

We could do something like this:
- before starting the app, check if `tdex-data` dir has anything in it. If it doesn't then this is a new app install. We create a file in the data directory called something like `HAS_MIGRATED` and then exit the script and allow the app to start.
- If `tdex-data` has data then check if the file called HAS_MIGRATED exists:
  - If it `HAS_MIGRATED` exists then we exit the script and allow the app to start normally. This will ensure that subsequent app restarts/updates don't nuke the database again.
  - If `HAS_MIGRATED` doesn't exist then we know that we need to migrate. We start the required container(s) for the migration tool and run the necessary commands to migrate. Then we create the `HAS_MIGRATED` file and exit the script so that the app starts up.

I can write this script on priority for you guys and test it out, I would just need guidance on the commands to run for migration. If I don't answer here, please feel free to DM me on X, Nostr, or Telegram!

**Comment by altafan** [comment]
@nmfretz the migration is an interactive process that requires the user to enter the password to unlock the data stored in the db. Can this be a problem for the workflow?

**Comment by tiero** [comment]
Guys can we merge this? People are still using the old deprecated version. 

Let's override the old data directory with bash?

**Comment by tiero** [comment]
Would maybe make an additional App ie TDEX V1  and deprecate this one be better? So we assume user do not expect a migration being a full restore

**Comment by nmfretz** [comment]
>Guys can we merge this? People are still using the old deprecated version.
>
>Let's override the old data directory with bash?

>Would maybe make an additional App ie TDEX V1 and deprecate this one be better? So we assume user do not expect a migration being a full restore

Hey @tiero. Alrighty, let's get this done. Without an automatic migration process on update, let's go with the following:

1) A warning at the top of the release notes explaining to users that this update will require them to restore the wallet from their seed phrase.

2) For existing installs we will keep the old data directory around so that in a worst-case scenario they can use this old database to restore using an older version of tdex. The way we can do this is just by simply changing the data directory bind mount on the host from say:
```
${APP_DATA_DIR}/tdex-data:/home/tdex/.tdex-daemon
```

to:

```
${APP_DATA_DIR}/tdexd:/home/tdex/.tdex-daemon
```

Let me make those changes and push them here so we can all test.

**Comment by nmfretz** [comment]
@tiero @altafan, here is the warning I have added to the top of the release notes. Does this look okay to you?

image

We could also include a link to your Telegram channel and say something like:

```yaml
If you encounter any issues when recovering your wallet after update, please reach out to the official TDEX Telegram group: https://t.me/tdexnetwork
```

If any users reach out with issues after updating, their old tdex daemon data directory will be located at:
`/home/umbrel/umbrel/app-data/tdex/tdex-data`

The new tdex daemon data directory is located at:
`/home/umbrel/umbrel/app-data/tdex/tdexd`

----------------------------

### Can you please take a look at this issue and see if you can recreate:

I just tested updating TDEX from v0.9.1 to v1.0.1. I went through the wallet recovery process and see this error in the UI:

image

I see this in the console:
image

But nothing in the networking tab, and nothing related in the container logs:

```
$ sudo docker logs -f tdex_tdexd_1
INFO[0000] All 0 tables opened in 0s
INFO[0000] Discard stats nextEmptySlot: 0
INFO[0000] Set nextTxnTs to 0
INFO[0000] All 0 tables opened in 0s
INFO[0000] Discard stats nextEmptySlot: 0
INFO[0000] Set nextTxnTs to 0
INFO[0000] All 0 tables opened in 0s
INFO[0000] Discard stats nextEmptySlot: 0
INFO[0000] Set nextTxnTs to 0
INFO[0000] All 0 tables opened in 0s
INFO[0000] Discard stats nextEmptySlot: 0
INFO[0000] Set nextTxnTs to 0
INFO[0000] All 0 tables opened in 0s
INFO[0000] Discard stats nextEmptySlot: 0
INFO[0000] Set nextTxnTs to 0
INFO[0000] starting daemon
INFO[0000] wallet interface is listening on :9092
DEBU[0008] /tdex_daemon.v2.WalletService/GetStatus
DEBU[0008] /tdex_daemon.v2.WalletService/GetInfo
DEBU[0016] /tdex_daemon.v2.WalletService/UnlockWallet
DEBU[0021] /tdex_daemon.v2.WalletService/GetStatus
DEBU[0021] /tdex_daemon.v2.WalletService/GetInfo
DEBU[0021] stopped server
INFO[0021] wallet interface is listening on :9092
INFO[0021] operator interface is listening on :9092
INFO[0021] trade interface is listening on :9092
DEBU[0044] /tdex_daemon.v2.WalletService/GetStatus
DEBU[0044] /tdex_daemon.v2.WalletService/GetInfo
DEBU[0063] /tdex_daemon.v2.WalletService/GetStatus
DEBU[0063] /tdex_daemon.v2.WalletService/GetInfo
DEBU[0600] Total allocated: 0.852GB, Heap allocated: 0.485GB, Allocated objects count: 1725134, Freed objects count: 1431569
DEBU[0600] Num of go routines: 102
```

```
$ sudo docker logs -f tdex_oceand_1
INFO[0000] All 1 tables opened in 0s
INFO[0000] Discard stats nextEmptySlot: 0
INFO[0000] Set nextTxnTs to 4
INFO[0000] All 0 tables opened in 0s
INFO[0000] Discard stats nextEmptySlot: 0
INFO[0000] Set nextTxnTs to 0
INFO[0000] All 0 tables opened in 0s
INFO[0000] Discard stats nextEmptySlot: 0
INFO[0000] Set nextTxnTs to 0
INFO[0000] service: registered wallet handler on public interface
INFO[0000] service: registered account handler on public interface
INFO[0000] service: registered transaction handler on public interface
INFO[0000] service: registered notification handler on public interface
DEBU[0000] scanner: start listening to messages from electrum server
INFO[0000] service: started blockchain scanner
INFO[0000] service: start listening on :18000
DEBU[0000] gRPC method: /ocean.v1.WalletService/Status
DEBU[0000] gRPC method: /ocean.v1.NotificationService/UtxosNotifications
DEBU[0000] gRPC method: /ocean.v1.NotificationService/TransactionNotifications
DEBU[0000] gRPC method: /ocean.v1.WalletService/GetInfo
DEBU[0009] gRPC method: /ocean.v1.WalletService/Status
DEBU[0009] gRPC method: /ocean.v1.WalletService/GetInfo
DEBU[0016] gRPC method: /ocean.v1.WalletService/Unlock
DEBU[0021] wallet repository: publish event WalletUnlocked
DEBU[0021] gRPC method: /ocean.v1.WalletService/GetInfo
DEBU[0021] gRPC method: /ocean.v1.WalletService/Status
DEBU[0022] gRPC method: /ocean.v1.WalletService/GetInfo
DEBU[0044] gRPC method: /ocean.v1.WalletService/Status
DEBU[0044] gRPC method: /ocean.v1.WalletService/GetInfo
DEBU[0064] gRPC method: /ocean.v1.WalletService/Status
DEBU[0064] gRPC method: /ocean.v1.WalletService/GetInfo
```

```
sudo docker logs -f tdex_dashboard_1
{"level":"info","ts":1718688563.9792867,"msg":"using provided configuration","config_file":"/etc/caddy/Caddyfile","config_adapter":"caddyfile"}
{"level":"info","ts":1718688563.9834294,"logger":"admin","msg":"admin endpoint started","address":"localhost:2019","enforce_origin":false,"origins":["//localhost:2019","//[::1]:2019","//127.0.0.1:2019"]}
{"level":"info","ts":1718688563.9839225,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0xc000764180"}
{"level":"info","ts":1718688563.9841962,"logger":"http.log","msg":"server running","name":"srv0","protocols":["h1","h2","h3"]}
{"level":"info","ts":1718688563.9845257,"msg":"autosaved config (load with --resume flag)","file":"/config/caddy/autosave.json"}
{"level":"info","ts":1718688563.9845374,"msg":"serving initial configuration"}
{"level":"info","ts":1718688563.984608,"logger":"tls","msg":"cleaning storage unit","description":"FileStorage:/data/caddy"}
{"level":"info","ts":1718688563.984663,"logger":"tls","msg":"finished cleaning storage units"}
```

**Comment by nmfretz** [comment]
Hey @altafan @tiero just pinging again in case you missed this comment here: https://github.com/getumbrel/umbrel-apps/pull/754#issuecomment-2175075416

**Comment by tiero** [comment]
@altafan @janaka-steph Can you please double-check this when you got time? To test with latest version (without involving migration of db)

**Comment by nmfretz** [comment]
@altafan @Janaka-Steph @tiero, just pinging in case this has fallen off your radar.

**Comment by Janaka-Steph** [comment]
I just tried to run tdex 1.0.1 in multipass VM. I encountered the same error "verification failed: signature mismatch after caveat verification". But after refreshing the page I was able to access the Tdex provider and load the wallet.

**Comment by nmfretz** [comment]
> I just tried to run tdex 1.0.1 in multipass VM. I encountered the same error "verification failed: signature mismatch after caveat verification". But after refreshing the page I was able to access the Tdex provider and load the wallet.
Feel free to merge, it is a minor issue.

Thanks @Janaka-Steph. Did you also try a fresh install and wallet creation where I encounter this:
image

**Comment by nmfretz** [comment]
@Janaka-Steph @tiero pinging again for the above. Thanks.

**Comment by Janaka-Steph** [comment]
@nmfretz I just tried and I didn't get error. It should have been fixed. Ok for merge!

**Comment by nmfretz** [comment]
Thanks @Janaka-Steph! Clearing local storage has gotten rid of this issue for me FYI, in case anyone reaches out with a similar issue. Going live.
